### PR TITLE
Adding source for code review

### PR DESCRIPTION
**Pull request by derekchiang**
@tsenart Please take a look.

### CODE REVIEW

File: cmd/scheduler/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
+package main
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: If we don't have many commands or separate library code, you should have this package in the repo root.

**Comment by tsenart** [comment]
Please run `go vet ./...`, `gofmt -s -w ./...` and `go lint` and address the warnings.

File: cmd/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+)
+
+const (
+	MemPerDaemonTask = 128 // mining shouldn't be memory-intensive
+	MemPerServerTask = 256 // I'm just guessing
+	CPUPerServerTask = 1   // a miner server does not use much CPU
+)
+
+// flags
+var (
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: In the interest of avoiding global state, define these flags inside main.

File: cmd/scheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+	master       = flag.String("master", "127.0.0.1:5050", "Master address <ip:port>")
+	bitcoindAddr = flag.String("bitcoindAddress", "127.0.0.1", "Address where bitcoind runs")
+
+	// auth
+	authProvider = flag.String("mesos_authentication_provider", sasl.ProviderName,
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Let's not mix snake and camel case flag names. For separators, I like to use dots.
`mesos.auth.provider`, `bitcoind.addr`

File: cmd/scheduler/scheduler.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+	tasksLaunched        int
+	currentDaemonTaskIDs []*mesos.TaskID
+}
+
+func newMinerScheduler(user, pass string) *MinerScheduler {
+	return &MinerScheduler{
+		rpcUser:              user,
+		rpcPass:              pass,
+		minerServerRunning:   false,
+		tasksLaunched:        0,
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: These two fields don't need initialization due to their zero-values.

File: cmd/scheduler/scheduler.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+		currentDaemonTaskIDs: make([]*mesos.TaskID, 0),
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: `[]*mesos.TaskID{}`

File: cmd/scheduler/scheduler.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+	}
+}
+
+func (s *MinerScheduler) Registered(driver sched.SchedulerDriver, frameworkId *mesos.FrameworkID, masterInfo *mesos.MasterInfo) {
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Unused arguments should be identified by an underscore: `_`

File: cmd/scheduler/scheduler.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+		for _, res := range cpuResources {
+			cpus += res.GetScalar().GetValue()
+		}
+
+		portsResources := util.FilterResources(offer.Resources, func(res *mesos.Resource) bool {
+			return res.GetName() == "ports"
+		})
+		var ports uint64
+		for _, res := range portsResources {
+			port_ranges := res.GetRanges().GetRange()
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: The convention is to use camel case, not snake case.

File: cmd/scheduler/scheduler.go [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+			}
+			log.Infof("Prepared task: %s with offer %s for launch\n", task.GetName(), offer.Id.GetValue())
+
+			tasks = append(tasks, task)
+			s.currentDaemonTaskIDs = append(s.currentDaemonTaskIDs, taskId)
+		}
+
+		driver.LaunchTasks([]*mesos.OfferID{offer.Id}, tasks, &mesos.Filters{RefuseSeconds: proto.Float64(1)})
+	}
+}
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: This method is 139 LOC. Please refactor it appropriately.

File: cmd/scheduler/scheduler.go [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+
+func (s *MinerScheduler) StatusUpdate(driver sched.SchedulerDriver, status *mesos.TaskStatus) {
+	log.Infoln("Status update: task", status.TaskId.GetValue(), " is in state ", status.State.Enum().String())
+	// If the mining server failed for any reason, kill all daemons, since they will be trying to talk to the failed mining server
+	if strings.Contains(status.GetTaskId().GetValue(), "server") &&
+		(status.GetState() == mesos.TaskState_TASK_LOST ||
+			status.GetState() == mesos.TaskState_TASK_KILLED ||
+			status.GetState() == mesos.TaskState_TASK_FINISHED ||
+			status.GetState() == mesos.TaskState_TASK_ERROR ||
+			status.GetState() == mesos.TaskState_TASK_FAILED) {
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Invert the conditional and return early to reduce nesting.

File: cmd/scheduler/scheduler.go [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+func (s *MinerScheduler) OfferRescinded(sched.SchedulerDriver, *mesos.OfferID) {}
+func (s *MinerScheduler) FrameworkMessage(sched.SchedulerDriver, *mesos.ExecutorID, *mesos.SlaveID, string) {
+}
+func (s *MinerScheduler) SlaveLost(sched.SchedulerDriver, *mesos.SlaveID) {}
+func (s *MinerScheduler) ExecutorLost(sched.SchedulerDriver, *mesos.ExecutorID, *mesos.SlaveID, int) {
+}
+
+func (sched *MinerScheduler) Error(driver sched.SchedulerDriver, err string) {
+}
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Add logging to these callbacks,

File: cmd/scheduler/scheduler.go [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+
+func printUsage() {
+	println("Usage: scheduler [--FLAGS] [RPC username] [RPC password]")
+	println("Your RPC username and password can be found in your bitcoin.conf file.")
+	println("To see a detailed description of the flags available, type `scheduler --help`")
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Don't use println.

File: cmd/scheduler/scheduler.go [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+func main() {
+	flag.Parse()
+
+	var user, pass string
+	if flag.NArg() == 1 {
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: switch would be more appropriate here.

File: cmd/scheduler/scheduler.go [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+		user = flag.Arg(0)
+		pass = flag.Arg(1)
+	} else {
+		printUsage()
+		println(flag.NArg())
+		return
+	}
+
+	fwinfo := &mesos.FrameworkInfo{
+		User: proto.String(""),
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: ?

File: cmd/scheduler/scheduler.go [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+		Name: proto.String("BTC Mining Framework (Go)"),
+	}
+
+	cred := (*mesos.Credential)(nil)
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: `var cred *mesos.Credential`

File: cmd/scheduler/scheduler.go [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+			Secret:    secret,
+		}
+	}
+	config := sched.DriverConfig{
+		Scheduler:  newMinerScheduler(user, pass),
+		Framework:  fwinfo,
+		Master:     *master,
+		Credential: cred,
+		WithAuthContext: func(ctx context.Context) context.Context {
+			ctx = auth.WithLoginProvider(ctx, *authProvider)
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Just: `return auth.WithLoginProvider(ctx, *authProvider)`

File: cmd/scheduler/scheduler.go [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+			return ctx
+		},
+	}
+
+	driver, err := sched.NewMesosSchedulerDriver(config)
+
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: Remove line.

File: cmd/scheduler/scheduler.go [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+	if err != nil {
+		log.Errorln("Unable to create a SchedulerDriver ", err.Error())
+	}
+
+	if stat, err := driver.Run(); err != nil {
+		log.Infof("Framework stopped with status %s and error: %s\n", stat.String(), err.Error())
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: New lines are automatically added to every log.[Method] invocation.

File: cmd/scheduler/scheduler.go [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+			s.minerServerPort = int(worker_port)
+
+			tasks = append(tasks, task)
+		}
+
+		if s.minerServerRunning && mems >= MemPerDaemonTask {
+			var taskId *mesos.TaskID
+			var task *mesos.TaskInfo
+
+			s.tasksLaunched += 1
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: `s.tasksLaunched++`

File: cmd/scheduler/scheduler.go [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+			taskId = &mesos.TaskID{
```
Comment (tsenart) [inline cmd/scheduler/scheduler.go]: `taskID`

**Comment by tsenart** [comment]
PTAL @derekchiang

**Comment by derekchiang** [comment]
@tsenart could you take another look?

File: README.md [hunk 20]
```diff
@@ -47,17 +47,17 @@ In the end, you should obtain the IP address and port where the Mesos master is
 You need to have [Go](http://golang.org/) installed.  Then in this directory:
 
     go get github.com/tools/godep # only if you don't have godep installed already
-    godep go build ./...
+    godep go build scheduler.go
```
Comment (tsenart) [inline README.md:50]: Why this change?

File: scheduler.go [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+			}
+			log.Infof("Prepared task: %s with offer %s for launch\n", task.GetName(), offer.Id.GetValue())
+
+			tasks = append(tasks, task)
+			s.currentDaemonTaskIDs = append(s.currentDaemonTaskIDs, taskID)
+		}
+
+		driver.LaunchTasks([]*mesos.OfferID{offer.Id}, tasks, &mesos.Filters{RefuseSeconds: proto.Float64(1)})
+	}
+}
```
Comment (tsenart) [inline scheduler.go:212]: 139 LOC is still too big! Factor things out into descriptive methods.

File: scheduler.go [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+Usage: scheduler [--FLAGS] [RPC username] [RPC password]
+Your RPC username and password can be found in your bitcoin.conf file.
+To see a detailed description of the flags available, type "scheduler --help"
+`)
+}
+
+func main() {
+
+	// flags
+	master := flag.String("master", "127.0.0.1:5050", "Master address <ip:port>")
```
Comment (tsenart) [inline scheduler.go:268]: Don't use global flag set since it is polluted by other packages. Instantiate your own.

File: README.md (hunk 20 above)
Comment (derekchiang) [inline README.md:50]: So the resulting binary is called `scheduler`.

File: scheduler.go (hunk 22 above)
Comment (derekchiang) [inline scheduler.go:268]: That was my intent in fact.  The `mesos-go` package adds some flags that the user might want to use.

File: scheduler.go (hunk 21 above)
Comment (tsenart) [inline scheduler.go:212]: OK. I assume you're fine with the state of this code.
